_posts/2021-01-XX-building-a-friendly-net-sdk-to-connect-to-jetbrains-space.markdown

```diff
@@ -0,0 +1,591 @@
+* `MessageRecipientChannel`
+* `MessageRecipientMember`
+* `MessageRecipientIssue`
+* ...
+
+You can use these in the .NET SDK for Space, like this:
+
+```csharp
+await chatClient.Messages.SendMessageAsync(
+    recipient: new MessageRecipientChannel()
+    {
+        Channel = new ChatChannelFromName()
+        {
+            Name = chatChannelName
+        }
+    }, // ....
+```
+
+This felt like a lot of ceremony, so we decided to let the code generator [create factory methods for inheritors](https://github.com/JetBrains/space-dotnet-sdk/blob/main/src/JetBrains.Space.Generator/CodeGeneration/CSharp/Generators/CSharpApiModelDtoGenerator.cs#L94). The above example now becomes this:
+
+```csharp
+await chatClient.Messages.SendMessageAsync(
+    recipient: MessageRecipient.Channel(ChatChannel.FromName(chatChannelName)), // ...
+```
+
+Shorter, less ceremony, and this has the nice side effect that [ReSharper](https://jetbrains.com/resharper) and [Rider](https://jetbrains.com/rider) prefer these in code completion, again helping with discoverability!
+
+![Rider code completion prefers factory methods](/images/2021/01/rider-code-completion.png)
+
+### More!
+
+If you're using ASP.NET Core and want to use Space as an authentication provider, have a look at the [`JetBrains.Space.AspNetCore.Authentication`](https://github.com/JetBrains/space-dotnet-sdk#jetbrainsspaceaspnetcoreauthentication) package. Sources are [here](https://github.com/JetBrains/space-dotnet-sdk/tree/main/src/JetBrains.Space.AspNetCore.Authentication). There's an interesting (experimental) API in there for [token management](https://github.com/JetBrains/space-dotnet-sdk#space-token-management-experimental), making it easy to use the Space API as the currently logged in user.
+
+Space supports building interactive applications, and we've added [experimental API's for those in the SDK](https://github.com/JetBrains/space-dotnet-sdk#space-applications-webhooks-experimental). Technically this is an abstraction built on top of a web API controller, but it again provides good developer experience. Here's an "echo" chat bot:
+
+```csharp
+public class CateringWebHookHandler : SpaceWebHookHandler
+{
+    private readonly ChatClient _chatClient;
+
+    public CateringWebHookHandler(ChatClient chatClient)
+    {
+        _chatClient = chatClient;
+    }
+
+    public override async Task HandleMessageAsync(MessagePayload payload)
+    {
+        if (payload.Message.Body is ChatMessageText messageText && !string.IsNullOrEmpty(messageText.Text))
+        {
+            await _chatClient.Messages.SendMessageAsync(
+                recipient: MessageRecipient.Channel(ChatChannel.FromId(payload.Message.ChannelId)),
+                content: ChatMessage.Text("You said: " + messageText.Text),
+                unfurlLinks: false);
+        }
+    }
+}
+```
+
+There are many more examples of focus on developer experience in the .NET SDK for Space, but since we've crossed 3500 words let's wrap up here...
```

Comment:
```suggestion
There are many more examples of this focus on developer experience in the .NET SDK for Space, but since we've crossed 3500 words let's wrap up here...
```